### PR TITLE
Issue/53 node search

### LINKED ISSUE #53: As a user, I can filter the roles/users I see in the role hierarchy vis by providing a case-insensitive search string and hitting filter in a filter menu

### PR DESCRIPTION
**Pull request by ganeshx**
- Provides functionality to search a node (User or a ROLE) in the network and see corresponding Objects filtered on the Object Hierarchy Panel on the right.

- Removes alerts for warning messages.

resolves #53, #83

### CODE REVIEW

File: src/js/app.js [hunk 1]
```diff
@@ -1,26 +1,31 @@
...
 import renderHierarchy from './hierarchy-vis';
+import '../../node_modules/jquery-ui-dist/jquery-ui';
+import '../../node_modules/jquery-ui-dist/jquery-ui.min.css';
 
-let data = {};
+// Needed for node search functionality
+let allNodeNames = [];
 
+let data = {};
 const init = (rawData) => {
```
Comment (randypitcherii) [inline src/js/app.js:11]: Please comment your function definitions

File: src/js/app.js [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
   data.nodesFiltered = false;
   data.formattedData = buildData(rawData);
   data.hierarchy = data.formattedData.hierarchy;
   data.renderedNetwork = renderNetwork(data.formattedData.nodes, data.formattedData.edges);
+  data.formattedData.nodes.map((x) => {
+    allNodeNames.push({ label: x.name, value: x.id });
+  });
   renderHierarchy(data.hierarchy);
 };
 const onNodeClicked = (currentNode, allChildren = []) => {
```
Comment (randypitcherii) [inline src/js/app.js:23]: Please comment your function definitions

File: src/js/app.js [hunk 3]
```diff
@@ -1,26 +1,31 @@
...
   if (currentNode.length > 0) {
-    if (!data.nodesFiltered) {
```
Comment (randypitcherii) [inline src/js/app.js:18]: add a comment explaining what you're doing here

File: src/js/app.js [hunk 4]
```diff
@@ -1,26 +1,31 @@
...
-      data.hierarchyFiltered = true;
-      const nodeArray = [];
-      nodeArray.push(currentNode[0]);
-      renderHierarchy(filterObjectsOnNodeClick(data.rawData, currentNode.concat(allChildren)));
-    }
+    data.hierarchyFiltered = true;
```
Comment (randypitcherii) [inline src/js/app.js:25]: comment to explain this

File: src/js/app.js [hunk 5]
```diff
@@ -32,18 +37,22 @@ const onNetworkReset = () => {
   renderHierarchy(data.hierarchy);
 };
 
-const isNetworkReset = () => {
+const isNodesFiltered = () => {
```
Comment (randypitcherii) [inline src/js/app.js:40]: Please comment your function definitions

File: src/js/app.js [hunk 6]
```diff
@@ -32,18 +37,22 @@ const onNetworkReset = () => {
...
   return data.nodesFiltered;
 };
 
+const isHierarchyFiltered = () => {
+  return data.hierarchyFiltered;
+};
+
 $('#hierarchy').on('select_node.jstree', function (e, jstreeObject) {
```
Comment (randypitcherii) [inline src/js/app.js:48]: Please comment your function definitions

File: src/js/app.js [hunk 7]
```diff
@@ -63,4 +72,44 @@ $('#hierarchy-vis').click(function (e) {
...
+    searchNode(ui.item.value);
+  },
+});
+
+const searchNode = (searchId) => {
+  data.renderedNetwork.resetNetwork();
+
+  let connectedNodeIds = [searchId];
+  let connectedNodes = data.renderedNetwork.showLinkedNodes(searchId);
+  connectedNodes.nodeArray.map((id) => {
```
Comment (randypitcherii) [inline src/js/app.js:91]: comment to explain this

File: src/js/data-builder.js [hunk 8]
```diff
@@ -12,21 +12,29 @@ const buildData = (json) => {
     rootType = {},
     dbRoot = {};
   json.forEach((element) => {
+    if (!element.toId) {
```
Comment (randypitcherii) [inline src/js/data-builder.js:15]: comment to explain this

File: src/js/data-builder.js [hunk 9]
```diff
@@ -188,9 +196,10 @@ const buildObjectNodeRelationship = (renderedNodes, filteredJson) => {
   let relatedData = [];
   filteredJson.forEach((element) => {
     if (element.GRANTED_TO_TYPE === 'ROLE' || element.GRANTED_TO_TYPE === 'USER') {
```
Comment (randypitcherii) [inline src/js/data-builder.js:198]: explain all of this with commenting

File: src/js/data-builder.js [hunk 10]
```diff
@@ -199,14 +208,12 @@ const buildObjectNodeRelationship = (renderedNodes, filteredJson) => {
   });
 
   //Return nodes that are related to provided filtered json
-  return renderedNodes.filter((node) => relatedData.some((item) => item.name === node.name));
+  return renderedNodes.filter((node) => relatedData.some((item) => item.id === node.id));
 };
 
 const filterObjectsOnNodeClick = (json, nodes = []) => {
```
Comment (randypitcherii) [inline src/js/data-builder.js:214]: Please comment your function definitions

File: src/js/data-builder.js [hunk 11]
```diff
@@ -199,14 +208,12 @@ const buildObjectNodeRelationship = (renderedNodes, filteredJson) => {
...
   let filteredJson = json.filter((element) => {
-    return nodes.filter((node) => node === element.GRANTED_TO_NAME && element.GRANTED_TO_TYPE === 'ROLE').length > 0
-      ? true
-      : false;
+    return nodes.filter((node) => node === element.toId).length > 0 ? true : false;
```
Comment (randypitcherii) [inline src/js/data-builder.js:216]: this ternary is redundant. The predicate will return the same values and the ternary makes this code harder to read

File: src/js/network-vis.js [hunk 12]
```diff
@@ -90,10 +90,10 @@ const renderNetwork = (dataNodes, dataEdges) => {
   // initialize network!
   let network = new Network(networkVisDiv, data, options);
 
-  const showFilteredNodes = (filterArray) => {
+  const showFilteredNodes = (filteredNodeIds) => {
```
Comment (randypitcherii) [inline src/js/network-vis.js:93]: Please comment your function definitions

File: src/js/network-vis.js [hunk 13]
```diff
@@ -90,10 +90,10 @@ const renderNetwork = (dataNodes, dataEdges) => {
...
     var view = new DataView(nodes, {
       filter: (item) => {
```
Comment (randypitcherii) [inline src/js/network-vis.js:95]: explain this with comments

File: src/js/network-vis.js [hunk 14]
```diff
@@ -202,7 +202,7 @@ const renderNetwork = (dataNodes, dataEdges) => {
   };
 
   const clearColorSelection = () => {
-    if (oldClickedNodeIds.length > 0) {
+    if (oldClickedNodeIds && oldClickedNodeIds.length > 0) {
```
Comment (randypitcherii) [inline src/js/network-vis.js:205]: explain this with comments

File: src/js/network-vis.js [hunk 15]
```diff
@@ -220,7 +220,7 @@ const renderNetwork = (dataNodes, dataEdges) => {
       });
       nodes.update(oldNodes);
     }
-    if (oldClickedEdgeIds.length > 0) {
+    if (oldClickedEdgeIds && oldClickedEdgeIds.length > 0) {
```
Comment (randypitcherii) [inline src/js/network-vis.js:223]: comments please

File: src/js/network-vis.js [hunk 16]
```diff
@@ -232,28 +232,38 @@ const renderNetwork = (dataNodes, dataEdges) => {
       });
       edges.update(oldClickedEdgeIds);
     }
+    oldClickedNodeIds = [];
```
Comment (randypitcherii) [inline src/js/network-vis.js:235]: explain this with comments please

File: src/js/network-vis.js [hunk 17]
```diff
@@ -232,28 +232,38 @@ const renderNetwork = (dataNodes, dataEdges) => {
...
+    oldClickedEdgeIds = [];
   };
 
-  const highlightNodes = (nodes) => {
-    clearColorSelection();
+  const highlightNodes = (nodes, clearSelection = true) => {
```
Comment (randypitcherii) [inline src/js/network-vis.js:239]: Please comment your function definitions

File: src/js/network-vis.js [hunk 18]
```diff
@@ -232,28 +232,38 @@ const renderNetwork = (dataNodes, dataEdges) => {
...
     let nodIds = [];
 
     nodes.map((node) => {
-      nodIds.push(node.name);
+      nodIds.push(node.id);
     });
     network.selectNodes(nodIds);
   };
+  //TODO: consolidate with network.on click handler
+  const nodeSelectionEventHandler = (nodeId, connectedNodes) => {
```
Comment (randypitcherii) [inline src/js/network-vis.js:249]: Please comment your function definitions

Also, please leave an empty space above your function definitions. This bunched up formatting is terrible to read

File: src/js/network-vis.js [hunk 19]
```diff
@@ -265,12 +275,15 @@ const renderNetwork = (dataNodes, dataEdges) => {
 
   network.on('click', (event) => {
     if (event.nodes.length === 0) {
-      if (event.edges.length === 0 && isNetworkReset()) {
+      if (event.edges.length === 0 && isNodesFiltered()) {
```
Comment (randypitcherii) [inline src/js/network-vis.js:278]: explain with comments please

File: src/js/new-visualization-modal.js [hunk 20]
```diff
@@ -2,41 +2,34 @@ import * as app from './app';
 
 import dataQuery from '../data/data_query.sql';
 
+let queryResults = null;
 // handle modal submit
 function onNewVisSubmit() {
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:7]: add blank space above all functions (obviously, above the comments. Keep the comments touching the function def)

File: src/js/new-visualization-modal.js [hunk 21]
```diff
@@ -2,41 +2,34 @@ import * as app from './app';
...
-
-  // plot the visualization
-  $('#create-vis-modal').modal('toggle');
   try {
-    app.init(JSON.parse(queryResults));
+    app.init(queryResults);
+    $('#create-vis-modal').modal('toggle');
   } catch (error) {
+    $('#invalid-query-results-warning').show();
     alert(
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:19]: remove alerts. Handle this without alerts

File: src/js/new-visualization-modal.js [hunk 22]
```diff
@@ -54,7 +47,54 @@ export default function initModal() {
...
   document.getElementById('data-query').innerHTML = dataQuery;
 
+  hideAllWarnings();
+
   // wire up event listeners
   $('#new-vis-button').click(onNewVisSubmit);
   $('#copy-query-button').click(onCopyQueryClick);
 }
+
+const csvStringToArray = (strData) => {
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:57]: Please comment your function definitions

File: src/js/new-visualization-modal.js [hunk 23]
```diff
@@ -54,7 +47,54 @@ export default function initModal() {
...
+  const objPattern = new RegExp('(\\,|\\r?\\n|\\r|^)(?:"([^"]*(?:""[^"]*)*)"|([^\\,\\r\\n]*))', 'gi');
+  // Create an array to hold our individual pattern
+  // matching groups.
+  let arrMatches = null,
+    // Create an array to hold our data. Give the array
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:62]: no reason to have such narrow comments. Collapse into single line comment please

Comment (randypitcherii) [reply src/js/new-visualization-modal.js:62]: do this for all similar comment breaks in this section

File: src/js/new-visualization-modal.js [hunk 24]
```diff
@@ -54,7 +47,54 @@ export default function initModal() {
...
+  // Keep looping over the regular expression matches
+  // until we can no longer find a match.
+  while ((arrMatches = objPattern.exec(strData))) {
+    if (arrMatches[1].length && arrMatches[1] !== '\\') arrData.push([]);
+    arrData[arrData.length - 1].push(arrMatches[2] ? arrMatches[2].replace(new RegExp('""', 'g'), '"') : arrMatches[3]);
+  }
+  return arrData;
+};
+
+function readSingleFile(evt) {
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:74]: Please comment your function definitions

File: src/js/new-visualization-modal.js [hunk 25]
```diff
@@ -54,7 +47,54 @@ export default function initModal() {
...
+  let f = evt.target.files[0];
+  if (f) {
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:76]: comment your if statements

File: src/js/new-visualization-modal.js [hunk 26]
```diff
@@ -54,7 +47,54 @@ export default function initModal() {
...
+      } catch (error) {
+        $('#invalid-query-results-warning').show();
+        alert(
+          "Whoops! We couldn't parse your results. Please double check your query and try again. Please submit feedback if you believe this is a bug."
+        );
+      }
+    };
+    r.readAsText(f);
+  } else {
+    alert('Failed to load file');
```
Comment (randypitcherii) [inline src/js/new-visualization-modal.js:97]: Do not use alerts. Find a better way to handle this issue.

**Comment by randypitcherii** [comment]
@ganeshx please do me a favor and resolve the conflicts as well. I accepted your PRs in order but there are still conflicts and I cannot resolve easily with the code in it's current state. Thank you.

File: src/js/app.js (hunk 1 above)
Comment (ganeshx) [reply src/js/app.js]: Self explanatory. function definition comment is noise than helpful in this case

File: src/js/app.js (hunk 2 above)
Comment (ganeshx) [reply src/js/app.js:27]: Done

File: src/js/app.js (hunk 3 above)
Comment (ganeshx) [reply src/js/app.js:19]: Done

File: src/js/app.js (hunk 4 above)
Comment (ganeshx) [reply src/js/app.js:29]: Done

File: src/js/app.js (hunk 5 above)
Comment (ganeshx) [reply src/js/app.js]: Self explanatory. Any further documentation is noise than helpful in this case

File: src/js/new-visualization-modal.js (hunk 22 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js:63]: Self explanatory.

File: src/js/new-visualization-modal.js (hunk 23 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js]: Done.

File: src/js/new-visualization-modal.js (hunk 24 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js]: Self explanatory. Any further documentation is noise than helpful

File: src/js/new-visualization-modal.js (hunk 25 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js]: Updated variable name for better reading. No need for comments in this case

File: src/js/new-visualization-modal.js (hunk 26 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js]: Alerts removed

File: src/js/app.js (hunk 6 above)
Comment (ganeshx) [reply src/js/app.js]: Self explanatory

File: src/js/app.js (hunk 7 above)
Comment (ganeshx) [reply src/js/app.js:98]: Done

File: src/js/data-builder.js (hunk 8 above)
Comment (ganeshx) [reply src/js/data-builder.js:16]: Done

File: src/js/data-builder.js (hunk 9 above)
Comment (ganeshx) [reply src/js/data-builder.js]: This file has a lot of business logic and needs a detailed wiki like documentation to explain the concepts of USER, ROLE, OBJECT, and their relationships. Or extensive code commenting and that need to be its own story as tech debt.

Please add a new story or as part of refactoring effort.

File: src/js/data-builder.js (hunk 10 above)
Comment (ganeshx) [reply src/js/data-builder.js]: Function definition is self explanatory.

File: src/js/data-builder.js (hunk 11 above)
Comment (ganeshx) [reply src/js/data-builder.js]: How is this redundant? if no node found, it return false. 

It's the opposite, for simple if else statements ternary operators are preferred. in fact react uses this heavily.

File: src/js/network-vis.js (hunk 12 above)
Comment (ganeshx) [reply src/js/network-vis.js:93]: self explanatory.

File: src/js/network-vis.js (hunk 13 above)
Comment (ganeshx) [reply src/js/network-vis.js:95]: self explanatory.

File: src/js/network-vis.js (hunk 14 above)
Comment (ganeshx) [reply src/js/network-vis.js:208]: Done

File: src/js/network-vis.js (hunk 16 above)
Comment (ganeshx) [reply src/js/network-vis.js:238]: Done

File: src/js/network-vis.js (hunk 18 above)
Comment (ganeshx) [reply src/js/network-vis.js:253]: Done

File: src/js/new-visualization-modal.js (hunk 20 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js:8]: It is followed for the most part. Some might have been overlooked. 
Fixed.

File: src/js/new-visualization-modal.js (hunk 21 above)
Comment (ganeshx) [reply src/js/new-visualization-modal.js]: Done

File: src/js/network-vis.js (hunk 19 above)
Comment (ganeshx) [reply src/js/network-vis.js]: Done

File: src/js/network-vis.js (hunk 17 above)
Comment (ganeshx) [reply src/js/network-vis.js:242]: self explanatory

File: src/js/network-vis.js (hunk 15 above)
Comment (ganeshx) [reply src/js/network-vis.js:226]: Don't see a point for comments on null checks.

**Comment by ganeshx** [comment]
I mentioned in a previous PR that I would no longer accept wildly uncommented PRs.

> I did not know about this. I had to go back and look at closed PRs one by one to see where you mentioned this. 
 Looks like you added review comments in PR #72 and then closed the PR immediately after that. I had no Ideas you had commented because I don't look at closed PRs unless there is a reason.

Please address the requested changes and use the change request as a pattern for future PRs. It's hard to apply exact rules for this, but here are some good guidelines for when you need a comment:

if you define a function. Maybe single line, well named functions don't need a comment but you should be commenting almost all your functions.

> Not needed for the most part unless we are developing a library (which is not the case here) as long as function name is self explanatory.

if you have a non human-readable if statement, please comment it
comment any regex to explain the pattern to a human
comment any complex state changes. If you reset a global variable, please explain what you're doing.
Lastly, I will never accept a PR without a description. Please ensure your PRs include:

description of the PR purpose (what feature are you adding or bug are you fixing)
list of high level changes
a link to the github issue you're addressing
add any notes about any special circumstances or new issues you've found.
Thank you for your hard work to add this code and I look forward to accepting it after you've made the requested changes.

**Pull request merged**